app/controllers/replies_controller.rb

```diff
@@ -297,19 +306,153 @@ def require_edit_permission
+    end
+    @multi_replies = @multi_replies_params.map do |reply_params|
+      Reply.new(reply_params).tap { |r| r.user = current_user }
+    end
+  end
+
+  def preview_reply(reply)
+    # Previewing a specific reply
+    @post = reply.post
+    @written = reply
     @written.user = current_user unless @written.user
     @audits = @written.id.present? ? { @written.id => @written.audits.count } : {}
+    @reply = @written # So that editor_setup shows the correct characters
+    preview_replies
+  end
 
+  def add_to_multi_reply(reply, reply_params)
+    # Adding a new reply to a multi reply
+    post_id = params[:reply][:post_id]
+    ReplyDraft.draft_for(post_id, current_user.id)&.destroy!
+
+    # Save NPC
+    if reply.character&.new_record?
+      if reply.character.save
+        flash[:success] = "Your new NPC has been persisted!"
+        params[:reply][:character_id] = reply.character.id
+        reply_params[:character_id] = reply.character.id
+        reply.character_id = reply.character.id
+      else
+        flash[:error] = "There was a problem persisting your new NPC."
+      end
+    end
+
+    # Set up editor
+    @post = reply.post
+    empty_reply_hash = permitted_params.permit(:character_id, :icon_id, :character_alias_id)
+    @empty_written = @post.build_new_reply_for(current_user, empty_reply_hash)
+    @empty_written.editor_mode = reply.editor_mode
+    @reply = @empty_written # So that editor_setup shows the correct characters
+    @audits = {}
+
+    # Add reply to list of multi replies
+    reply.user = current_user unless reply.user
+    @multi_replies << reply
+    reply_params[:id] = reply.id if reply.id.present?
+    @multi_replies_params << reply_params
+
+    preview_replies
+  end
+
+  def preview_replies
     @page_title = @post.subject
 
-    @reply = @written # So that editor_setup shows the correct characters
     editor_setup
     render :preview
   end
 
+  def post_replies(new_reply: nil)
+    @multi_replies << new_reply if new_reply.present?
+
+    first_reply = @multi_replies.first
+    begin
+      Reply.transaction { @multi_replies.each(&:save!) }
+    rescue ActiveRecord::RecordInvalid => e
+      errored_reply = @multi_replies.detect { |r| r.errors.present? } || first_reply
+      render_errors(errored_reply, action: 'created', now: true, err: e)
+
+      redirect_to posts_path and return unless errored_reply.post
+      redirect_to post_path(errored_reply.post) and return
+    end
+
+    if @multi_replies.length == 1
+      flash[:success] = "Reply posted."
+    else
+      flash[:success] = "Replies posted."
+    end
+    redirect_to reply_path(first_reply, anchor: "reply-#{first_reply.id}")
+  end
+
+  def editing_multi_reply?
+    # If the list of params is present and the first item on the list has the ID stored, I am editing it
+    @multi_replies_params.present? && (@reply = Reply.find_by_id(@multi_replies_params.first["id"]))
```

Comment:
```suggestion
    @multi_replies_params.present? && (@reply = Reply.find_by(id: @multi_replies_params.first["id"]))
```

Reply:
I'm pretty sure we're using find_by_id most elsewhere in the code?

Reply:
I thought we were moving away from find_by_id to find_by but -- this is probably also a subject for discord